### ISSUE TITLE
Add extensibility features

### ISSUE THREAD

**Issue by pkarw** [DivanteLtd/storefront-query-builder#1]
It would be awesome if the users can extend the way we build queries (eg. adding new operators and transformations). The easiest way would be to give the developers a way to inject their own filter transformers/mappers to the: [buildQueryBodyFromSearchQuery](https://github.com/DivanteLtd/storefront-query-builder/blob/7ec02de5fb6d8040fcd22d6f6049698be241e178/src/elasticsearch/index.ts#L67)

**Comment by pkarw** [comment]
Related task: https://github.com/DivanteLtd/vue-storefront/issues/3934

**Comment by cewald** [comment]
I think the current state is already giving an abstract and simple way to create one dimensional queries. It would sure be nice to extend this set of filters by custom or recurring ones if needed.
But where should they be registered – in the `/api/catalog` endpoint by something like the API's `ProductProcessor` using the configs?

This way we could apply a filter like in the VSF and add a `custom_filter` in the API to handle it:
```
searchQuery.applyFilter({ key: 'custom_filter', value: {  /* More complex stuff here */ } })
```

But then again, aren't we just need a way to run specific and complex DSL queries directly – without a new syntax that needs to be understandable and complex enough for developers? I see the agnostic benefit by using a own syntax and methods.

If I think about use-cases for extensibility, I think about an complex "select this, but not that, containing this" alike query for showing e.g: only yellow or blue shirts, containing a specific brand, but only if the desired size is not null, etc. . To make things like this possible we would end up rebuilding a `bodybuilder`-like library but ideally platform agnostic – am I wrong?

Long story short ;) — Why not give an opportunity to extend the default `SearchQuery` as described for "simple" or recurring queries extensions but also allow plain DSL objects for complex queries. This might be not platform agnostic – but you probably won't switch between DB systems without testing your most complex queries. We then also could handle them separately in the language specific platform of the query-builder like we already do for `elasticsearch`.

**Comment by cewald** [comment]
Some progress for that in: #6
